src/navigation/navigation_main.cc

```diff
@@ -180,44 +189,51 @@ void OdometryCallback(const nav_msgs::Odometry& msg) {
   navigation_.UpdateOdometry(odom_);
 }
 
-void LaserHandler(const sensor_msgs::LaserScan& msg) {
+void LaserHandler(const sensor_msgs::LaserScan& msg,
+                  const string& topic) {
   if (FLAGS_v > 2) {
-    printf("Laser t=%f, dt=%f\n",
+    printf("Laser topic=%s, t=%f, dt=%f\n",
+           topic.c_str(),
            msg.header.stamp.toSec(),
            GetWallTime() - msg.header.stamp.toSec());
   }
-  // Location of the laser on the robot. Assumes the laser is forward-facing.
-  const Vector2f kLaserLoc(CONFIG_laser_loc_x, CONFIG_laser_loc_x);
-  static float cached_dtheta_ = 0;
-  static float cached_angle_min_ = 0;
-  static size_t cached_num_rays_ = 0;
-  static vector<Vector2f> cached_rays_;
-  if (cached_angle_min_ != msg.angle_min ||
-      cached_dtheta_ != msg.angle_increment ||
-      cached_num_rays_ != msg.ranges.size()) {
-    cached_angle_min_ = msg.angle_min;
-    cached_dtheta_ = msg.angle_increment;
-    cached_num_rays_ = msg.ranges.size();
-    cached_rays_.resize(cached_num_rays_);
-    for (size_t i = 0; i < cached_num_rays_; ++i) {
-      const float a =
-          cached_angle_min_ + static_cast<float>(i) * cached_dtheta_;
-      cached_rays_[i] = Vector2f(cos(a), sin(a));
+
+  static unordered_map<string, LaserCache> laser_caches_;
+  laser_caches_.emplace(topic, LaserCache());
+  LaserCache& cache = laser_caches_[topic];
+
+  if (cache.dtheta != msg.angle_increment ||
+      cache.angle_min != msg.angle_min ||
+      cache.rays.size() != msg.ranges.size()) {
+    cache.dtheta = msg.angle_increment;
+    cache.angle_min = msg.angle_min;
+    cache.rays.resize(msg.ranges.size());
+    for (size_t i = 0; i < cache.rays.size(); ++i) {
+      const float a = cache.angle_min + static_cast<float>(i) * cache.dtheta;
+      cache.rays[i] = Vector2f(cos(a), sin(a));
     }
   }
-  CHECK_EQ(cached_rays_.size(), msg.ranges.size());
-  point_cloud_.resize(cached_rays_.size());
-  for (size_t i = 0; i < cached_num_rays_; ++i) {
+  CHECK_EQ(cache.rays.size(), msg.ranges.size());
+
+  size_t start_idx = point_cloud_.size();
+  point_cloud_.resize(start_idx + cache.rays.size());
+  // Location of the laser on the robot. Assumes the laser is forward-facing.
+  const Vector2f kLaserLoc(CONFIG_laser_loc_x, CONFIG_laser_loc_y);
```

Comment:
This won't work for multiple topics. Let's print a warning if the offset is non-zero, and multiple laser topics are subscribed.

Reply:
Should we remove the offset and verify if the `frame_id` of messages from multiple topics is set to `base_link` (defined in a configuration file), issuing a warning if so?"

Reply:
I push the changes